src/pools/D3MAaveDaiPool.sol

```diff
@@ -0,0 +1,161 @@
+    function claimRewards(address[] calldata assets, uint256 amount, address to) external returns (uint256);
+}
+
+contract D3MAaveDaiPool is D3MPoolBase {
+
+    uint256 constant RAY  = 10 ** 27;
+
+    address                  public immutable pool;
+    RewardsClaimerLike       public immutable rewardsClaimer;
+    ShareTokenLike           public immutable stableDebt;
+    ShareTokenLike           public immutable variableDebt;
+    address                  public immutable interestStrategy;
+    address                  public immutable adai; // Token representing a share of the asset pool
+
+    address public king;  // Who gets the rewards
+
+    event Collect(address indexed king, address[] assets, uint256 amt);
+    event File(bytes32 indexed what, address data);
+
+    constructor(address hub_, address dai_, address pool_, address _rewardsClaimer) public D3MPoolBase(hub_, dai_) {
+        pool = pool_;
+
+        // Fetch the reserve data from Aave
+        (,,,,,,, address adai_, address stableDebt_, address variableDebt_, address interestStrategy_,) = LendingPoolLike(pool_).getReserveData(dai_);
+        require(adai_ != address(0), "D3MAaveDaiPool/invalid-adai");
+        require(stableDebt_ != address(0), "D3MAaveDaiPool/invalid-stableDebt");
+        require(variableDebt_ != address(0), "D3MAaveDaiPool/invalid-variableDebt");
+        require(interestStrategy_ != address(0), "D3MAaveDaiPool/invalid-interestStrategy");
+
+        adai = adai_;
+        stableDebt = ShareTokenLike(stableDebt_);
+        variableDebt = ShareTokenLike(variableDebt_);
+        interestStrategy = interestStrategy_;
+        rewardsClaimer = RewardsClaimerLike(_rewardsClaimer);
+
+        ShareTokenLike(adai_).approve(pool_, type(uint256).max);
+        TokenLike(dai_).approve(pool_, type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "D3MAaveDaiPool/overflow");
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Admin ---
+    function file(bytes32 what, address data) external auth {
+        require(live == 1, "D3MTestPool/no-file-not-live");
+
+        if (what == "king") king = data;
+        else revert("D3MPoolBase/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function validTarget() external view override returns (bool) {
+        (,,,,,,,,,, address strategy,) = LendingPoolLike(pool).getReserveData(address(asset));
+        return strategy == interestStrategy;
+    }
```

Comment:
I'm still not a big fan of this. The interest strategy is not relevant to the pool and even if that goes to something entirely new it's not a big deal as we are limited by debt ceilings. We can just use a governance vote to shut it down.

Furthermore `validTarget()` without any args is not super generalizable, so I would prefer it not be a part of the interface. See https://github.com/makerdao/dss-crop-join/blob/master/src/SushiJoinV2.sol#L152 as an example where you need to specify args to front-run a timelock change coming you don't like. I would prefer to keep this omitted entirely. It doesn't help us much and seems like it won't be used elsewhere.

Reply:
Hmm I guess that can be solved creating a function in the `pool` that expects the random params and sets a storage value which makes `validTarget()` to return `false`.
What I might agree is that the name might be more generic. Maybe just make the pool to have an `active` flag and use that for `Hub.cage`, or maybe using the actual `Pool.live`. In that case `Pool.cage` can implement the custom things you were mentioning.

Reply:
So in general, I think this is a helpful function to have.  Yes we could respond with governance action, but we are not guaranteed to be able to take that action in time (i.e. if our delay is greater than theirs).

For the sushi piece, what would stop us from creating an internal function in a custom pool that provides those? Nothing needs to come from the hub for that, right?

I like Gonza's suggestion on cage, but of course I just removed all of that from the pool based on other comments.  

Bottom line for me: I think if a function is called from the hub on a pool, it should be in the base/interface.  I think this particular function provides some protection for us beyond what we can get with governance actions and debt ceilings.  If we decide to remove it, I would probably lean towards a pool cage function that also blocks more things in the pool (like deposit/withdraw) to give us the most amount of protection.

Reply:
Fair point on some custom function to trip a `live` or something like that for the sushiswap case.

I guess to me this particular trip-wire doesn't make sense anymore. It's so specific and pool doesn't even use the interest strategy (separation of concerns). If there was one or two things we could put defensive actions on then sure, but Aave governance can literally upgrade the contract to whatever they want, so I don't see what this gives us.

Even if they upgraded the interest rate strategy without letting us know worst case it would just have us potentially depositing too much DAI or reverting in which case we would just cage it manually a few days later.

I think a better way to handle this trip-wire situation is have Plan's able to have custom `disable(...)` which will set the target debt ceiling to 0. So in this case there could be a permissionless `disable()` which checks the interest rate strategy in Plan and sets the `bar` to 0. I still don't think it's all that useful here, but if you really want to keep it I think it belongs in the Plan and not the Pool.

Reply:
Ok yes, I see the point that checking the interestStrategy makes more sense in the plan than in the pool.
The only caveat is the Plan will need to cache the Pool as well in order to do this check (speaking always about the specific Aave case).

What I'm wondering if we should just leave a flag in the pool and in the plan which if any of them are true we can cage the collateral in the `Hub`. e.g. a flag called `cage`.
So then in `Hub.cage(ilk)` we will check if `Pool.cage or Plan.cage`. That leaves us a very general solution to implement whatever permissionless thing we want/need in the plans and pools.

Reply:
I'd be okay with this live check on both pool and plan. I still don't think this adds much protection with the interest rate strategy changing though. :)

Reply:
> Ok yes, I see the point that checking the interestStrategy makes more sense in the plan than in the pool. The only caveat is the Plan will need to cache the Pool as well in order to do this check (speaking always about the specific Aave case).
> 
> What I'm wondering if we should just leave a flag in the pool and in the plan which if any of them are true we can cage the collateral in the `Hub`. e.g. a flag called `cage`. So then in `Hub.cage(ilk)` we will check if `Pool.cage or Plan.cage`. That leaves us a very general solution to implement whatever permissionless thing we want/need in the plans and pools.

@gbalabasquer do you think we should have `cage` (i.e. one way street to kill something) or a disable like we have in the plan right now that could be reversed by governance? I am leaning towards disable so we could turn off a d3m temporarily, unwind it and figure out next steps.  What do you think?

Reply:
This is what I have for now without the cage/hub stuff: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/837685402efb489532acc33ebea9cb75cf5cd50d

Reply:
I added the active checks here: https://github.com/makerdao/dss-direct-deposit/pull/31/files/974a0de8213635f399d9af79770a3e6bcbee41ab..894134a1d7ddd2d9d11f00b5159ee7fc8ffb32c6

I re-integrated them with the exec function (basically we will unwind the positions if either is not active).

However, I did not add them back to the `cage(ilk)` function.  My thinking is two fold:
1. Right now we are only caging the pool (in the Hub) so it wasn't exactly clear to me that we should cage only the pool or that it should happen if the plan is disabled.
2. In general, my concept of `active`/`disable` is more temporary. Basically a safety measure for us to be able to pause a D3M and put ourselves in a safe position while we figure things out.  Because `cage` is one way, I felt that was a little heavy handed for this. 

Let me know your thoughts on this approach @talbaneth @gbalabasquer and @hexonaut

Reply:
The `strategy != address(interestStrategy)` check should be in aave plan's `active()` function, so it will take affect instantly.

If we want the disable to be temporary I think you can make `interestStrategy` updateable (and add a function for that) so governance can resume the d3m pool operation without need to redeploy and unwind the old funds over time if the pool is empty.

Reply:
We have tried to keep the `pool` and the `plan` from knowing about each other, so I don't want to introduce that entanglement right now. Also, we could have many `pool`s using the same `plan`.  I'm not sure what you mean by "instantly".

I was talking about the whole D3M concept being temporarily disabled.  I was thinking that if the Aave interest strategy changed we would temporarily disable the plan, unwind the D3M position. We would then launch a new plan based on the new strategy and could file it in the Hub and and rewind up the system.  I don't feel too strongly about filing the interest strategy but since it shouldn't change too often, probably fine to just swap the plan when it does change.  Let me know if you think we should have the file and I'll add that.

Reply:
Tal suggests to avoid having a `disable` function in the `AavePlan` and instead making:
```
function active() external view override returns (bool) {
return strategy == address(interestStrategy);
}
```

That way you can protect 100% of the contract being changed without needing a keeper to interact in the middle.

Reply:
Actually that was the main reason to add the `active` checks in the `exec` function itself.

Reply:
Thank you! i misread `plan` and `pool` (multiple times... 😬 )

So does that mean we can/should get rid of the disable function all together?